base/c.jl

```diff
@@ -503,3 +503,183 @@ end
+struct CcallError{T <: AbstractString} <: Exception
+    msg::T
+end
+Base.showerror(io::IO, e::CcallError) = print(io, "CcallError: ", e.msg)
+
+getcfunc(f) = QuoteNode(f)
+getcfunc(f::Expr) = :(($(f.args[2]), $(f.args[1])))
+hashead(symbol, _) = false
+hashead(expr::Expr, head) = expr.head === head
+
+islinum(node) = node isa LineNumberNode
+
+remove_linums(e::Expr) =
+    Expr(e.head, remove_linums.(filter(!islinum, e.args))...)
+remove_linums(s) = s
+
+"""
+Determine if there are varargs and return a vector of all arguments
+based on the call signature.
+
+returns a tuple of `(arguments, varargs)`
+"""
+function getargs(call)
+    # no arguments
+    length(call.args) < 2 &&
+        return [], []
+
+    firstarg = call.args[2]
+    hashead(firstarg, :parameters) &&
+        return call.args[3:end], firstarg.args
+
+    return call.args[2:end], []
+end
+
+"""
+       mkarg(arg)
+
+`mkarg` takes an argument and returns a tuple of symbol and type. The
+argument must have a type annotation or an error will be thrown.
+"""
+function mkarg(arg)
+    !hashead(arg, :(::)) &&
+        throw(CcallError("args in @ccall need type annotations. '$arg' doesn't have one."))
+    return (arg=arg.args[1], type=arg.args[2])
+end
+
+"""
+get the type specified on the varargs. throw if there is more than one type.
+"""
+function getvarargtype(varargs)
+    vararg_types = Set(a.type for a in varargs)
+    length(vararg_types) > 1 &&
+        throw(CcallError("varargs @ccall with different argument types not yet supported"))
+    return pop!(vararg_types)
+end
+
+"""
+    parsecall(expression)
+
+`parsecall` is an implementation detail of `@ccall
+
+it takes and expression like `:(printf("%d"::Cstring, value::Cuint)::Cvoid)`
+returns: a tuple of `(function_name, return_type, arg_types, args)`
+
+The above input outputs this:
+
+    (:printf, :Cvoid, :((Cstring, Cuint)), ["%d", :value])
+
+Note that the args are in an array, not a quote block and have to be
+appended to the ccall in a separate step.
+"""
+function parsecall(expr::Expr)
+    # setup and check for errors
+    !hashead(expr, :(::)) && begin
+        throw(CcallError("@ccall needs a function signature with a return type"))
+    end
+    rettype = expr.args[2]
+
+    call = expr.args[1]
+    !hashead(call, :call) &&
+        throw(CcallError("@ccall has to take a function call"))
+
+    # get the function symbols
+    func = getcfunc(call.args[1])
+    normalargs, varargs = getargs(call)
+
+
+    # separate annotations from names
+    normalargs = mkarg.(normalargs)
+    args = Any[a.arg for a in normalargs]
+    types = Any[a.type for a in normalargs]
+    argtypes = :(())
+    argtypes.args = types
+
+    isempty(varargs) && return func, rettype, argtypes, args, 0
+    nreq = length(args)
+
+    # vararg handling.
+    varargs = mkarg.(varargs)
+    append!(args, (a.arg for a in varargs))
+    append!(argtypes.args, (a.type for a in varargs))
+    return func, rettype, argtypes, args, nreq
```

Comment:
Might be even more idiomatic (for the fairly small set of parsers-written-in-Julia) to make `pusharg!` a closure.

Reply:
This function was getting kind of hairy and was in need of much love. "more imperative" isn't something I usually strive for in my code, but if that's what the Julia maintainers prefer, that's fine. mostly copy-pasted this in https://github.com/JuliaLang/julia/pull/32748/commits/bbd6c5af06e54f25126de65d9e0cae7c0d4cd846 with a few tweaks.

One thing to note is that I leave `nreq` at 0 if there are no varargs because this is what `ccall` lowers to. I don't know if that's correct. I just copied what I saw because I know it's not broken. Maybe @vtjnash knows something about this topic.

Reply:
That's because `nreq == 0` is illegal in C. Well, except for this case:
```C
$ clang -x c -emit-llvm -S - -o -
int func();
int main() {
return func();
}
```
```llvm
declare i32 @func(...)
define i32 @main() {
  %1 = call i32 (...) @func()
  ret i32 %1
}
```
But since we're passing zero arguments, that's probably OK not to have specific syntax for this in Julia.

Reply:
> "more imperative" isn't something I usually strive for in my code

Haha, well said! I guess my overall thought was that the flow of this function seemed to have gotten a little confused during the refactoring, but describing exactly what I mean by this is not straightforward. I think the existing version of this code with the `append!` invocations and early return error handling etc was quite imperative in style, so my thought was that we may as well be unapologetic about it ;-)